### DISCUSSION TITLE
Finding the effective key definition

### DISCUSSION THREAD

**Discussion by raducoravu** [dita-ot/dita-ot#4231] [Q&A] [answered]
I'm attaching a sample DITA Map, it has a main "main_map.ditamap" and a topic named "subordinate-maps/topics/test_topic.dita".
The "test_topic.dita" has a key reference to a key named "product":

    <xref keyref="product"/>

My question is, based on the DITA specification what should be the resolved value of the keyref? Should the keyref resolve to "product1A.dita" or to "product2.dita"?

The specs says:

     Put another way, a key definition is effective if it is the first definition for that key name in the shallowest map that contains that key definition.

https://www.oxygenxml.com/dita/1.3/specs/archSpec/base/keys-core-concepts.html#concept_igf_3c4_kn__effective-key-definition-full

[effective-key-def.zip](https://github.com/dita-ot/dita-ot/files/12014611/effective-key-def.zip)

@drmacro it would be great if you could give me some input on this.

**Comment by ekimbernow** [comment, marked as answer]
Based on my initial inspection of the map, the effective value of the key "product" as referenced by test_topic.dita is "product2" because that definition is first in the breadth-first traversal of the map tree. 

The definition of the key "product" in sub_map1A.ditamap is in a child of the peer map of sub_map2.ditamap, which makes it a lower priority. So even though the definition in sub_map1A.ditamap is earlier in a depth-first traversal of the resolved map, it is encountered later in a breadth-first traversal.

What I see in Oxygen for the resolved value in test_topic.dita looks correct to me.

**Comment by lief-erickson** [comment]
@raducoravu, my understanding is that it should resolve to "product2.dita". The reason is because "product2" is defined in a submap, but "product1A" is defined in a sub-submap, making it deeper in the hierarchy than the shallower "product2." If you were to define `@keys="product"` in the main map, then it would be the shallowest and would win, even if the definition is after all of the maprefs and topicrefs.

If two key definitions are found at the same level, then first one wins (again, it's "shallower"). You can test this in your files by adding `<keydef keys="product" href="topics/product1A.dita"/>` in `sub_map1.ditamap`.

Imagine the logic being something like this, with each level being resolved before moving on to the next and with duplicates at the same level are resolved with found-first logic:

| | Depth |
|-|-|
|Are there any key definitions in the main map? |Shallowest|
|Are there any key definitions in any submaps?| Deeper|
|Are there any key definitions in any sub-submaps?| Deeper yet|
|Are there any key definitions in any sub-sub-submaps?| Even deeper still| 
|Are there any key definitions in any sub-sub-sub-submaps? |Least shallow / deepest|

I think this table maybe explains @drmacro's "breadth-first traversal".

**Comment by raducoravu** [comment]
@ekimbernow @lief-erickson thanks, I agree with your analysis.
The DITA OT 4.1 publishing engine resolves the "product" keyref to "product1A.html". So I will open an issue for this.

**Comment by ekimbernow** [comment]
If it's any consolation, using your test data set, I found the same bug in my XQuery key space construction code.
